addon/ember-concurrency.ts

```diff
@@ -0,0 +1,99 @@
+import { task as _task, taskGroup as _taskGroup } from 'ember-concurrency';
+import RSVP from 'rsvp';
+
+export const task = _task as (fn: () => IterableIterator<any>) => TaskProperty;
+
+export const taskGroup = _taskGroup as () => TaskGroupProperty;
+
+interface CommonTaskProperty {
+  restartable: () => TaskProperty;
+  drop: () => TaskProperty;
+  keepLatest: () => TaskProperty;
+  enqueue: () => TaskProperty;
+  maxConcurrency: (n: number) => TaskProperty;
+  cancelOn: (eventName: string) => TaskProperty;
+  group: (groupName: string) => TaskProperty;
+}
+
+export interface TaskProperty extends CommonTaskProperty {
+  evented: () => TaskProperty;
+  debug: () => TaskProperty;
+  on: (eventName: string) => TaskProperty;
+}
+
+export interface TaskGroupProperty extends CommonTaskProperty {}
+
+export type TaskOptions = {
+  [option in keyof TaskProperty]?: Parameters<TaskProperty[option]> extends [
+    infer P
+  ]
+    ? P
+    : true
+};
+export type TaskGroupOptions = {
+  [option in keyof TaskGroupProperty]?: Parameters<
+    TaskGroupProperty[option]
+  > extends [infer P]
+    ? P
+    : true
+};
+
+// Based on https://github.com/CenterForOpenScience/ember-osf-web/blob/7933316efae805e00723789809bdeb58a96a286a/types/ember-concurrency/index.d.ts
+
+export enum TaskInstanceState {
+  Dropped = 'dropped',
+  Canceled = 'canceled',
+  Finished = 'finished',
+  Running = 'running',
+  Waiting = 'waiting'
+}
+
+export interface TaskInstance<T> extends PromiseLike<T> {
+  readonly error?: any;
+  readonly hasStarted: boolean;
+  readonly isCanceled: boolean;
+  readonly isDropped: boolean;
+  readonly isError: boolean;
+  readonly isFinished: boolean;
+  readonly isRunning: boolean;
+  readonly isSuccessful: boolean;
+  readonly state: TaskInstanceState;
+  readonly value?: T;
+  cancel(): void;
+  catch(): RSVP.Promise<any>;
+  finally(): RSVP.Promise<any>;
+  then<TResult1 = T, TResult2 = never>(
+    onfulfilled?:
+      | ((value: T) => TResult1 | RSVP.Promise<TResult1>)
+      | undefined
+      | null,
+    onrejected?:
+      | ((reason: any) => TResult2 | PromiseLike<TResult2>)
+      | undefined
+      | null
+  ): RSVP.Promise<TResult1 | TResult2>;
```

Comment:
`RSVP.Promise<T>` is a `class` that brings along a `constructor` and `static` methods. I _guess_ that this is the reason it wasn't done like this originally. Maybe @chriskrycho can enlighten us?

Reply:
Chris is on vacation, but I think I can explain this one.

When you use `class`, it creates both a value _and_ an interface of the same name. The type contains everything you'd see on an instance of the class, and the value contains all the static members of that class, as well as a `prototype` field and a `new` signature corresponding to the class's `constructor`. In other words:

```ts
declare class Foo1 {
    static prop: string;
    constructor(arg: number);
    method(): number;
}

/* 👆 is roughly equivalent to 👇 */

interface Foo2 {
    method(): number;
}

declare const Foo2: {
    prop: string;
    new(arg: number): Foo2;
    prototype: Foo2;
}
```

For normal development of actual runtime code, this tends to be an interesting but mostly irrelevant detail. When you're working with type declarations, though, it's super useful, because it means each `class` actually produces an `interface` to represent the type of its instances that's not bound to the mechanical details of the constructor function itself.

It so happens that [the only non-`static` members of `RSVP.Promise` are `then`, `catch` and `finally`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/rsvp/index.d.ts#L113-L124), so saying our interface extends that one is essentially a shorthand for saying it has those three methods.

Another alternative would be something like:

```ts
export interface TaskInstance<T> {
  // ...
  then: typeof RSVP.Promise['then'];
  catch: typeof RSVP.Promise['catch'];
  finally: typeof RSVP.Promise['finally'];
}
```